File: Library/PaintedSkin/src/main/java/org/alee/component/skin/executor/ListViewSkinExecutor.java
package org.alee.component.skin.executor;

import static org.alee.component.skin.parser.DefaultExecutorBuilder.ATTRIBUTE_LIST_VIEW_DIVIDER;

import android.content.res.ColorStateList;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.ColorStateListDrawable;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.widget.ListView;

import androidx.annotation.NonNull;

/**********************************************************
 *
 * @author: MY.Liu
 * @date: 2021/2/13
 * @description: xxxx
 *
 *********************************************************/
public class ListViewSkinExecutor<T extends ListView> extends ViewSkinExecutor<T> {


    public ListViewSkinExecutor(@NonNull SkinElement fullElement) {
        super(fullElement);
    }

    @Override
    protected void applyColor(@NonNull T view, @NonNull ColorStateList colorStateList, @NonNull String attrName) {
        super.applyColor(view, colorStateList, attrName);
        switch (attrName) {
            case ATTRIBUTE_LIST_VIEW_DIVIDER:
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    applyDrawable(view, new ColorStateListDrawable(colorStateList), attrName);
                } else {
                    applyColor(view, colorStateList.getDefaultColor(), attrName);
                }
                break;
            default:
                break;
        }
    }

    @Override
    protected void applyColor(@NonNull T view, int color, @NonNull String attrName) {
        super.applyColor(view, color, attrName);
        switch (attrName) {
            case ATTRIBUTE_LIST_VIEW_DIVIDER:
                applyDrawable(view, new ColorDrawable(color), attrName);
                break;
            default:
                break;
        }
    }

    @Override
    protected void applyDrawable(@NonNull T view, @NonNull Drawable drawable, @NonNull String attrName) {
        super.applyDrawable(view, drawable, attrName);
        switch (attrName) {
            case ATTRIBUTE_LIST_VIEW_DIVIDER:
                int dvh = view.getDividerHeight();
                view.setDivider(drawable);
                view.setDividerHeight(dvh);
                break;
            default:
                break;
        }
    }
}

File: Library/PaintedSkin-Kotlin/src/main/java/org/alee/component/skin/core/executor/ListViewSkinExecutor.kt
package org.alee.component.skin.core.executor

import android.content.res.ColorStateList
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.ColorStateListDrawable
import android.graphics.drawable.Drawable
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import android.widget.ListView
import org.alee.component.skin.constant.DefaultSkinAttribute.AbsListViewAttribute
import org.alee.component.skin.constant.DefaultSkinAttribute.ListViewAttribute
import org.alee.component.skin.template.ISkinExecutor

/**
 * [ListView] 属性换肤执行器
 *
 * <p> 所有[ListView]及其子类View的[ISkinExecutor]都应该继承于此类
 * <p> 实现了对以下属性的换肤：
 * <p> [AbsListViewAttribute.ATTRIBUTE_CACHE_COLOR_HINT]
 * <p> [AbsListViewAttribute.ATTRIBUTE_LIST_SELECTOR]
 *
 * @author MingYu.Liu
 * created in 2023/9/8
 *
 */
open class ListViewSkinExecutor<T : ListView> : AbsListViewSkinExecutor<T>() {

    override fun applyColor(view: T, colorStateList: ColorStateList, attrName: String) {
        super.applyColor(view, colorStateList, attrName)
        when (attrName) {
            ListViewAttribute.ATTRIBUTE_LIST_VIEW_DIVIDER -> {
                if (VERSION.SDK_INT >= VERSION_CODES.Q) {
                    applyDrawable(view, ColorStateListDrawable(colorStateList), attrName)
                } else {
                    applyColor(view, colorStateList.defaultColor, attrName)
                }
            }
        }
    }

    override fun applyColor(view: T, color: Int, attrName: String) {
        super.applyColor(view, color, attrName)
        when (attrName) {
            ListViewAttribute.ATTRIBUTE_LIST_VIEW_DIVIDER -> {
                applyDrawable(view, ColorDrawable(color), attrName)
            }
        }
    }

    override fun applyDrawable(view: T, drawable: Drawable, attrName: String) {
        super.applyDrawable(view, drawable, attrName)
        when (attrName) {
            ListViewAttribute.ATTRIBUTE_LIST_VIEW_DIVIDER -> {
                val dvh = view.dividerHeight
                view.divider = drawable
                view.dividerHeight = dvh
            }
        }
    }
}
